datumaro/plugins/synthetic_data/image_generator.py

```diff
@@ -0,0 +1,268 @@
+        if not (
+            osp.exists(proto_path) and osp.exists(model_path) and osp.exists(hull_path)
+        ) and not os.access(save_dir, os.W_OK):
+            raise ValueError(
+                "Please provide a path to a colorization model directory or "
+                "a path to a writable directory to download the model"
+            )
+
+        for url, filename, size, md5_checksum in [
+            (
+                f"https://raw.githubusercontent.com/richzhang/colorization/a1642d6ac6fc80fe08885edba34c166da09465f6/colorization/models/{prototxt_file_name}",
+                prototxt_file_name,
+                9945,
+                "7229f469d24645a7f1e3c47c67e7bd15",
+            ),
+            (
+                f"http://eecs.berkeley.edu/~rich.zhang/projects/2016_colorization/files/demo_v2/{caffemodel_file_name}",
+                caffemodel_file_name,
+                128946764,
+                "6ef9d30cefd880baabeb5e1847fb20be",
+            ),
+            (
+                f"https://raw.githubusercontent.com/richzhang/colorization/a1642d6ac6fc80fe08885edba34c166da09465f6/colorization/resources/{hull_file_name}",
+                hull_file_name,
+                5088,
+                "05c2729d850b5e4143b6fe53326066b5",
+            ),
+        ]:
+            save_path = osp.join(save_dir, filename)
+            if osp.exists(save_path):
+                continue
+
+            log.info("Downloading the '%s' file to '%s'", filename, save_dir)
+            try:
+                cls._download_file(
+                    url, save_path, expected_size=size, expected_checksum=md5_checksum
+                )
+            except Exception as e:
+                raise Exception(f"Failed to download the '{filename}' file: {str(e)}")
+
+    @staticmethod
+    @scoped
+    def _download_file(
+        url: str, output_path: str, *, timeout: int = 60, expected_size: int, expected_checksum: str
+    ) -> None:
+        BLOCK_SIZE = 2**20
+
+        response = requests.get(url, timeout=timeout, stream=True)
+        response.raise_for_status()
+
+        checksum_counter = hashlib.md5()  # nosec B303 - not a security context
+        actual_size = 0
+
+        on_error_do(os.unlink, output_path)
+
+        with open(output_path, "wb") as fd:
+            for chunk in response.iter_content(chunk_size=BLOCK_SIZE):
+                actual_size += len(chunk)
+                if actual_size > expected_size:
+                    # There is also the context-length header, but it can be corrupted or invalid
+                    # for different reasons
+                    raise Exception(
+                        f"The downloaded file has unexpected size {actual_size}. "
+                        f"Expected {expected_size}."
+                    )
+
+                checksum_counter.update(chunk)
+
+                fd.write(chunk)
+
+        actual_checksum = checksum_counter.hexdigest()
+        if actual_checksum != expected_checksum:
```

Comment:
It is better to compare digests as binary strings rather than hex strings, so that differences in representation (e.g. character case) don't affect the result.